lib/mix/lib/mix/compilers/test.ex

```diff
@@ -45,24 +45,43 @@ defmodule Mix.Compilers.Test do
         warnings_as_errors? = Keyword.get(opts, :warnings_as_errors, false)
 
         try do
-          failed? =
+          compilation_failed? =
             case Kernel.ParallelCompiler.require(test_files, parallel_require_callbacks) do
-              {:ok, _, [_ | _]} when warnings_as_errors? -> true
-              {:ok, _, _} -> false
-              {:error, _, _} -> exit({:shutdown, 1})
+              {:ok, modules, [_ | _] = warnings} when warnings_as_errors? ->
+                # IO.inspect(second, label: "second")
+                files_with_warnings = Enum.map(warnings, &elem(&1, 0))
+                # IO.inspect(warnings, label: "warnings")
+                # IO.inspect(files_with_warnings, label: "files_with_warnings")
+                :ok  =
+                  ExUnit.mark_files_as_failed(
+                    modules,
+                    files_with_warnings,
+                    opts[:failures_manifest_file]
+                  )
```

Comment:
To avoid the race condition, you need to do this after the suite runs (after `await_run`). Also, is `failures_manifest_file` guaranteed to exist? Or can the user set it to nil somehow?

Reply:
After another look, I am afraid this still won't be enough. The warning might have come from a support file and no test will be marked as dirty.

I am starting to think the best option is to indeed check on the status code and mark this as unresolved. :(

Reply:
Yeah, given all the options, the simplest/cleanest solution seems to be the one I suggested.
WDYT about adding a note on this issue on the help text for `--warnings-as-errors` and/or `--failed`?